cpp/include/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,489 @@
+}
+
+namespace detail {
+
+    // `exponent` comes from using scale_type = weak_typedef<int32_t>
+    //  we still need Rep to know what type we want for accumulator of ipow
+    template <typename Rep, Radix Base, typename T,
+              typename std::enable_if_t<(std::is_same<int32_t, T>::value
+                                      && is_supported_representation_type<Rep>())>* = nullptr>
+    Rep ipow(T exponent) {
+        // Integer Exponentiation by Squaring:
+        // https://simple.wikipedia.org/wiki/Exponentiation_by_squaring
+        // Note: this is the iterative equivalent of the recursive definition (faster)
+        // Quick-bench: http://quick-bench.com/Wg7o7HYQC9FW5M0CO0wQAjSwP_Y
+        if (exponent == 0) return static_cast<Rep>(Base);
+        auto extra  = static_cast<Rep>(1);
+        auto square = static_cast<Rep>(Base);
+        while (exponent > 1) {
+            if (exponent & 1 /* odd */) {
+                extra *= square;
+                exponent -= 1;
+            }
+            exponent /= 2;
+            square *= square;
+        }
+        return square * extra;
+    }
+
+    // helper function to negate strongly typed scale_type
+    CUDA_HOST_DEVICE_CALLABLE
+    auto negate(scale_type const& scale) {
+        return scale_type{-scale};
+    }
+
+    // perform this operation when constructing with positive scale
+    template <typename Rep, Radix Rad, typename T>
+    CUDA_HOST_DEVICE_CALLABLE
+    constexpr T right_shift(T const& val, scale_type const& scale) {
+        assert(scale > 0);
+        return val / ipow<Rep, Rad>(scale._t);
+    }
+
+    // perform this operation when constructing with negative scale
+    template <typename Rep, Radix Rad, typename T>
+    CUDA_HOST_DEVICE_CALLABLE
+    constexpr T left_shift(T const& val, scale_type const& scale) {
+        assert(scale < 0);
+        return val * ipow<Rep, Rad>(-scale._t);
+    }
+
+    // convenience generic shift function
+    template <typename Rep, Radix Rad, typename T>
+    CUDA_HOST_DEVICE_CALLABLE
+    constexpr T shift(T const& val, scale_type const& scale) {
+        if      (scale == 0) return val;
+        else if (scale >  0) return right_shift<Rep, Rad>(val, scale);
+        else                 return left_shift <Rep, Rad>(val, scale);
+    }
+
+    template <typename Rep, Radix Rad, typename T,
+              typename std::enable_if_t<is_supported_construction_value_type<T>()>* = nullptr>
+    auto shift_with_precise_round(T const& value, scale_type const& scale) -> int64_t {
+        if (scale == 0) return value;
+        int64_t const base   = static_cast<int64_t>(Rad);
+        int64_t const factor = ipow<int64_t, Rad>(std::abs(scale));
+        int64_t const temp   = scale <= 0 ? value * (factor * base) : value / (factor / base);
+        return std::roundf(static_cast<double>(temp) / base);
+    }
+
+}
+
+/**---------------------------------------------------------------------------*
```

Comment:
We've gradually been removing all these horizontal lines in docs. 

```suggestion
/**
```

Reply:
Done.